### PR TITLE
Add docker

### PR DESCRIPTION
**Pull request by Yolnan**

### CODE REVIEW

**Comment by Yolnan** [comment]
@marip8 When entering the snp_automate_2022 container, I get notified that "groups: cannot find name for group ID ..." and my username is "I have no name!". I made sure to set the CURRENT_UID environment variable.

Do we want to add a user to the Dockerfile?  It currently seems to be root.

**Comment by marip8** [comment]
> @marip8 When entering the snp_automate_2022 container, I get notified that "groups: cannot find name for group ID ..." and my username is "I have no name!". I made sure to set the CURRENT_UID environment variable.

Is your machine get your user name from the domain? One of mine does and I get the same error; my other machine is not on a domain and I don't run into this issue. In any case, this issue shouldn't prevent your from running anything

FWIW, mounting the directories below let's the docker image know where your user name and password information can be found

```
      - /etc/passwd:/etc/passwd:ro
      - /etc/group:/etc/group:ro
```

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    paths:
+      - 'snp**'
+      - 'docker/Dockerfile'
+      - '.github/workflows/docker.yml'
```
Comment (marip8) [inline .github/workflows/docker.yml]: Remove
```suggestion
```

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            type=ref,event=pr,prefix=${{ matrix.distro }}-
+            type=semver,pattern={{major}}.{{minor}},prefix=${{ matrix.distro }}-
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: docker/Dockerfile
+          build-args: |
+            TAG=${{ matrix.distro }}-0.21
```
Comment (marip8) [inline .github/workflows/docker.yml]: The SNP docker images only have the suffix `-master` at the moment, no version numbers

```suggestion
            TAG=${{ matrix.distro }}-master
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+USER root
+
+# Install Python dependencies
+RUN python3 -m pip install open3d pyquaternion --upgrade
+
+# Bind mount the source directory so as not to unnecessarily copy source code into the docker image
+ARG WORKSPACE_DIR=/opt/snp_automate_2022
+RUN --mount=type=bind,target=${WORKSPACE_DIR}/src/snp_automate_2022 \
+  apt update -y -qq \
+  && source /opt/snp/install/setup.bash \
```
Comment (marip8) [inline docker/Dockerfile:18]: Let's also install some useful utilities like `sudo` for the image

```suggestion
# Run apt update and install useful utilities like sudo
RUN apt update -y -qq && apt install -y -qq sudo

# Bind mount the source directory so as not to unnecessarily copy source code into the docker image
ARG WORKSPACE_DIR=/opt/snp_automate_2022
RUN --mount=type=bind,target=${WORKSPACE_DIR}/src/snp_automate_2022 \
  && source /opt/snp/install/setup.bash \
```

**Review by marip8 — CHANGES_REQUESTED** [review]

**Comment by Yolnan** [comment]
> > @marip8 When entering the snp_automate_2022 container, I get notified that "groups: cannot find name for group ID ..." and my username is "I have no name!". I made sure to set the CURRENT_UID environment variable.
> 
> Is your machine get your user name from the domain? One of mine does and I get the same error; my other machine is not on a domain and I don't run into this issue. In any case, this issue shouldn't prevent your from running anything
> 
> FWIW, mounting the directories below let's the docker image know where your user name and password information can be found
> 
> ```
>       - /etc/passwd:/etc/passwd:ro
>       - /etc/group:/etc/group:ro
> ```

I think my laptop is. I did test the container and it seems to work despite the username. I have those directories in the docker compose.yml. Oddly, I think Tyler recommended that I remove them.

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+FROM ghcr.io/ros-industrial-consortium/scan_n_plan_workshop:${TAG}
+
+SHELL ["/bin/bash", "-c"]
+
+ENV DEBIAN_FRONTEND noninteractivej
+
+USER root
+
+# Install Python dependencies
+RUN python3 -m pip install open3d pyquaternion --upgrade
```
Comment (marip8) [inline docker/Dockerfile]: Remove; comes from SNP docker image now

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+# Bind mount the source directory so as not to unnecessarily copy source code into the docker image
+ARG WORKSPACE_DIR=/opt/snp_automate_2022
+RUN --mount=type=bind,target=${WORKSPACE_DIR}/src/snp_automate_2022 \
+  apt update -y -qq \
+  && source /opt/snp/install/setup.bash \
+  && vcs import ${WORKSPACE_DIR}/src < ${WORKSPACE_DIR}/src/snp_automate_2022/dependencies.repos --shallow \
+  && rosdep install \
+    --from-paths ${WORKSPACE_DIR}/src \
+    -iry \
+    --skip-keys libvtk
```
Comment (marip8) [inline docker/Dockerfile]:
```suggestion
    -iry
```

File: docker/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# SNP Automate 2022 Docker
+
+## Create the Docker image
+### Download
+Download a pre-built Docker image from the container registry:
+
+```
+docker login ghcr.io
+docker pull ghcr.io/ros-industrial-consortium/scan_n_plan_workshop:<tag>
```
Comment (marip8) [inline docker/README.md:9]:
```suggestion
docker pull ghcr.io/ros-industrial-consortium/snp_automate_2022:<tag>
```

File: docker/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    build:
+      context: ..
+      dockerfile: docker/Dockerfile
+      args:
+        - TAG=foxy-master
+    environment:
+      DISPLAY: $DISPLAY
+      XAUTHORITY: $XAUTHORITY
+      NVIDIA_DRIVER_CAPABILITIES: all
+      ROS_LOG_DIR: /tmp
```
Comment (marip8) [inline docker/docker-compose.yml]:
```suggestion
      ROS_LOG_DIR: /tmp
      ROS_LOCALHOST_ONLY: 1
```

File: docker/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    stdin_open: true
+    tty: true
+    network_mode: host
+    privileged: true
+    user: ${CURRENT_UID}  # CURRENT_UID=$(id -u):$(id -g)
+    volumes:
+      - /tmp/.X11-unix:/tmp/.X11-unix
+      - /etc/hosts:/etc/hosts
+      - /etc/passwd:/etc/passwd:ro
+      - /etc/group:/etc/group:ro
```
Comment (marip8) [inline docker/docker-compose.yml:29]: Add `/dev` for USB devices

```suggestion
      - /etc/group:/etc/group:ro
      - /etc/shadow:/etc/shadow:ro
      - /dev:/dev
```

File: docker/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    deploy:
+      resources:
+        reservations:
+          devices:
+            - driver: nvidia
+              count: 1
+              capabilities: [gpu]
```
Comment (marip8) [inline docker/docker-compose.yml]: Add the ROS1 bridge

```suggestion
              capabilities: [gpu]
  # ROS1 bridge
  ros1-bridge:
    environment:
      ROS_LOG_DIR: /tmp
      ROS_LOCAL_HOST_ONLY: 1
    container_name: ros1-bridge
    image: ghcr.io/ros-industrial/ros1_bridge:noetic-foxy
    tty: true
    privileged: true
    network_mode: host
```

**Review by marip8 — CHANGES_REQUESTED** [review]

**Review by marip8 — APPROVED** [review]

**Pull request merged**